src/plugins/data/public/search/aggs/agg_config.ts

```diff
@@ -281,7 +294,61 @@ export class AggConfig {
       enabled: this.enabled,
       type: this.type && this.type.name,
       schema: this.schema,
-      params: outParams,
+      params: outParams as SerializableState,
+    };
+  }
+
+  /**
+   * @deprecated - Use serialize() instead.
+   */
+  toJSON(): AggConfigSerialized {
+    return this.serialize();
+  }
+
+  /**
+   * @returns Returns an ExpressionAst representing the function for this agg type.
+   */
+  toExpressionAst(): ExpressionAstExpression | undefined {
+    const functionName = this.type && this.type.expressionName;
+    const { type, ...rest } = this.serialize();
+    if (!functionName || !rest.params) {
+      // Return undefined - there is no matching expression function for this agg
+      return;
+    }
+
+    // Go through each of the params and convert to an array of expression args.
+    const params = Object.entries(rest.params).reduce((acc, [key, value]) => {
+      const deserializedParam = this.getAggParams().find(p => p.name === key);
+
+      if (deserializedParam && deserializedParam.type === 'agg') {
+        // We can call `toExpressionAst` if our param is another agg
+        acc[key] = [this.getParam(key).toExpressionAst()];
+      } else if (typeof value === 'object') {
+        // For some params which are object, like ranges, we stringify for now
+        acc[key] = [JSON.stringify(value)];
+      } else if (typeof value !== 'undefined') {
+        // Everything else just gets stored in an array if it is defined
+        acc[key] = [value];
+      }
+
+      return acc;
+    }, {} as Record<string, ExpressionAstArgument[]>);
+
+    return {
+      type: 'expression',
+      chain: [
+        {
+          type: 'function',
+          function: functionName,
+          arguments: {
+            ...params,
+            // Expression args which are provided to all functions
+            id: [this.id],
+            enabled: [this.enabled],
+            ...(this.schema ? { schema: [this.schema] } : {}), // schema may be undefined
+          },
```

Comment:
The only weird thing here is that toExpressionAst returns the AST as an entire expression, not just as the AST for an individual function.

This was necessary because of params of type `agg`, on which we call `.toExpressionAst()` as well. Since the AST for a subexpression must be an entire expression AST (not just another function), it means this function needs to return the whole AST expression.

This creates a small amount of extra work for consumers, as they will need to manually take the value of `ast.chain[0]` to include the function in their expression. This felt like an okay tradeoff, but open to better ideas.

Reply:
Fixed this in the latest update; it returns `ExpressionAstFunction` now